src/System.Private.CoreLib/shared/System/Collections/ObjectModel/Collection.cs

```diff
@@ -162,6 +170,76 @@ protected virtual void SetItem(int index, T item)
             items[index] = item;
         }
 
+        protected virtual void InsertItemsRange(int index, IEnumerable<T> collection)
+        {
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            if (GetType() == typeof(Collection<T>) && items is List<T> list)
+            {
+                list.InsertRange(index, collection);
+            }
+            else
+            {
+                foreach (T item in collection)
+                {
+                    InsertItem(index++, item);
+                }
+            }
+        }
+
+        protected virtual void RemoveItemsRange(int index, int count)
+        {
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+            }
+
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            if (index > items.Count - count)
+            {
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
+            }
+
+            int length = items.Count;
+            if (index < items.Count - count)
+            {
+                length = index + count;
+            }
+
+            for (int i = index; i < length; i++)
```

Comment:
This should now just be:
```C#
for (int i = 0; i < count; i++)
{
   RemoveItem(index);
}
```

Reply:
What is the expectation of the `ReplaceItemsRange` API. Currently we have several test cases that started failing after making this change:

ReplateItemsRange:
```c#
protected virtual void RemoveItemsRange(int index, int count)
{
    if (items.IsReadOnly)
    {
        ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
    }

    if ((uint)index > (uint)items.Count)
    {
        ThrowHelper.ThrowArgumentOutOfRange_IndexException();
    }

    if (count < 0)
    {
        ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
    }

    if (index > items.Count - count)
    {
        ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
    }            

    for (int i = index; i < count; i++)
    {
        RemoveItem(index);
    }
}
```

Here is the code from one of the failing test cases
```c#
[Fact]
public void Collection_ReplaceRange_LastTwo_Test()
{
    int[] initial = new int[] { 1, 2, 3, 4 };
    int[] replace = new int[] { 5, 6, 7, 8 };
    Collection<int> collection = new Collection<int>();
    foreach (var item in initial)
        collection.Add(item);

    collection.ReplaceRange(2, 2, replace);

    Assert.NotNull(collection);
    Assert.Equal(initial.Length + 2, collection.Count);

    int[] collectionAssertion = collection.ToArray();
    Assert.Equal(initial.AsSpan(0, 2).ToArray(), collectionAssertion.AsSpan(0, 2).ToArray());
    Assert.Equal(replace.AsSpan(0, 4).ToArray(), collectionAssertion.AsSpan(2, 4).ToArray());
}
```

While the API states it takes the following parameters:
* index - where to start the replace
* count - the number of items to delete
* collection - the collection to insert into the initial at position index

The [spec](https://github.com/dotnet/corefx/issues/10752#issuecomment-427452499) where @safern stated:
> Will allow to replace a range with fewer, equal, or more items.

Given the unit test mentioned above the final result will be
`[ 1, 2, 5, 6, 7, 8 ]`

With the proposed change documented in this comment above where we specify
`for (int i = index; i < count; i++)`

I believe this needs to be updated to
`for (int i = index; i < count + index; i++)`

After testing this change all the unit tests are passing but, it is making the assumption that the entire collection will be inserted even it is larger than the count specified.

## My Question
Is this business rule correct?
> Will allow to replace a range with fewer, equal, or more items.

Or should the replace API be constrained to the count

Reply:
`for (int i = index; i < count; i++)`

That's not what I suggested though.

Start at 0, not index.

This: 
`for (int i = index; i < count + index; i++)`

Is the same as:
`for (int i = 0; i < count; i++)`

Reply:
I'm really sorry, I'm dyslexic and it usually doesn't effect me this bad in PRs. 

Thank you for spelling this out so clearly for me and this makes 100% sense to me. Update is on it's way after I run through the tests

Reply:
> I'm really sorry, I'm dyslexic and it usually doesn't effect me this bad in PRs.

No worries. I appreciate your responsiveness and drive for clarity :)

Keep going!

Reply:
Thanks for the kind words! 

Updated the PR with latest changes. I'm going to mark this as resolved.